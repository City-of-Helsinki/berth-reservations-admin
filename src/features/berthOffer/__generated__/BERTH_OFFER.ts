/* tslint:disable */
/* eslint-disable */
// @generated
// This file was automatically generated and should not be edited.

import { ApplicationStatus, LeaseStatus, BerthMooringType } from "./../../../@types/__generated__/globalTypes";

// ====================================================
// GraphQL query operation: BERTH_OFFER
// ====================================================

export interface BERTH_OFFER_berthApplication_berthSwitch {
  __typename: "BerthSwitchType";
  id: string;
}

export interface BERTH_OFFER_berthApplication_customer {
  __typename: "ProfileNode";
  id: string;
}

export interface BERTH_OFFER_berthApplication {
  __typename: "BerthApplicationNode";
  id: string;
  createdAt: any;
  status: ApplicationStatus;
  berthSwitch: BERTH_OFFER_berthApplication_berthSwitch | null;
  customer: BERTH_OFFER_berthApplication_customer | null;
  boatType: string | null;
  boatRegistrationNumber: string | null;
  boatName: string | null;
  boatModel: string | null;
  boatWidth: any | null;
  boatLength: any | null;
  boatDraught: any | null;
  boatWeight: number | null;
}

export interface BERTH_OFFER_boatTypes {
  __typename: "BoatTypeType";
  id: string;
  name: string | null;
}

export interface BERTH_OFFER_harbor_properties_piers_edges_node_properties_berths_edges_node_leases_edges_node_customer {
  __typename: "ProfileNode";
  id: string;
}

export interface BERTH_OFFER_harbor_properties_piers_edges_node_properties_berths_edges_node_leases_edges_node {
  __typename: "BerthLeaseNode";
  status: LeaseStatus;
  startDate: any;
  endDate: any;
  isActive: boolean;
  customer: BERTH_OFFER_harbor_properties_piers_edges_node_properties_berths_edges_node_leases_edges_node_customer;
}

export interface BERTH_OFFER_harbor_properties_piers_edges_node_properties_berths_edges_node_leases_edges {
  __typename: "BerthLeaseNodeEdge";
  node: BERTH_OFFER_harbor_properties_piers_edges_node_properties_berths_edges_node_leases_edges_node | null;
}

export interface BERTH_OFFER_harbor_properties_piers_edges_node_properties_berths_edges_node_leases {
  __typename: "BerthLeaseNodeConnection";
  edges: (BERTH_OFFER_harbor_properties_piers_edges_node_properties_berths_edges_node_leases_edges | null)[];
}

export interface BERTH_OFFER_harbor_properties_piers_edges_node_properties_berths_edges_node {
  __typename: "BerthNode";
  comment: string;
  depth: number | null;
  id: string;
  isAccessible: boolean | null;
  isActive: boolean;
  leases: BERTH_OFFER_harbor_properties_piers_edges_node_properties_berths_edges_node_leases | null;
  length: number;
  mooringType: BerthMooringType;
  number: string;
  width: number;
}

export interface BERTH_OFFER_harbor_properties_piers_edges_node_properties_berths_edges {
  __typename: "BerthNodeEdge";
  node: BERTH_OFFER_harbor_properties_piers_edges_node_properties_berths_edges_node | null;
}

export interface BERTH_OFFER_harbor_properties_piers_edges_node_properties_berths {
  __typename: "BerthNodeConnection";
  edges: (BERTH_OFFER_harbor_properties_piers_edges_node_properties_berths_edges | null)[];
}

export interface BERTH_OFFER_harbor_properties_piers_edges_node_properties {
  __typename: "PierProperties";
  identifier: string;
  electricity: boolean;
  gate: boolean;
  water: boolean;
  lighting: boolean;
  wasteCollection: boolean;
  berths: BERTH_OFFER_harbor_properties_piers_edges_node_properties_berths;
}

export interface BERTH_OFFER_harbor_properties_piers_edges_node {
  __typename: "PierNode";
  id: string;
  properties: BERTH_OFFER_harbor_properties_piers_edges_node_properties | null;
}

export interface BERTH_OFFER_harbor_properties_piers_edges {
  __typename: "PierNodeEdge";
  node: BERTH_OFFER_harbor_properties_piers_edges_node | null;
}

export interface BERTH_OFFER_harbor_properties_piers {
  __typename: "PierNodeConnection";
  edges: (BERTH_OFFER_harbor_properties_piers_edges | null)[];
}

export interface BERTH_OFFER_harbor_properties {
  __typename: "HarborProperties";
  name: string | null;
  servicemapId: string | null;
  imageFile: string | null;
  streetAddress: string | null;
  municipality: string | null;
  zipCode: string;
  maxWidth: number | null;
  numberOfPlaces: number;
  numberOfFreePlaces: number;
  piers: BERTH_OFFER_harbor_properties_piers | null;
}

export interface BERTH_OFFER_harbor {
  __typename: "HarborNode";
  id: string;
  properties: BERTH_OFFER_harbor_properties | null;
}

export interface BERTH_OFFER {
  berthApplication: BERTH_OFFER_berthApplication | null;
  boatTypes: BERTH_OFFER_boatTypes[] | null;
  harbor: BERTH_OFFER_harbor | null;
}

export interface BERTH_OFFERVariables {
  applicationId: string;
  harborId: string;
}
