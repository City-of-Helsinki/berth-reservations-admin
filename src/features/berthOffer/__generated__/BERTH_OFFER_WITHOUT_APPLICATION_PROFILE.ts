/* tslint:disable */
/* eslint-disable */
// @generated
// This file was automatically generated and should not be edited.

// ====================================================
// GraphQL query operation: BERTH_OFFER_WITHOUT_APPLICATION_PROFILE
// ====================================================

export interface BERTH_OFFER_WITHOUT_APPLICATION_PROFILE_profile_boats_edges_node_boatType {
  __typename: "BoatTypeType";
  id: string;
  name: string | null;
}

export interface BERTH_OFFER_WITHOUT_APPLICATION_PROFILE_profile_boats_edges_node {
  __typename: "BoatNode";
  id: string;
  boatType: BERTH_OFFER_WITHOUT_APPLICATION_PROFILE_profile_boats_edges_node_boatType;
  length: any;
  width: any;
  draught: any | null;
  weight: number | null;
  name: string;
  model: string;
  registrationNumber: string;
}

export interface BERTH_OFFER_WITHOUT_APPLICATION_PROFILE_profile_boats_edges {
  __typename: "BoatNodeEdge";
  node: BERTH_OFFER_WITHOUT_APPLICATION_PROFILE_profile_boats_edges_node | null;
}

export interface BERTH_OFFER_WITHOUT_APPLICATION_PROFILE_profile_boats {
  __typename: "BoatNodeConnection";
  edges: (BERTH_OFFER_WITHOUT_APPLICATION_PROFILE_profile_boats_edges | null)[];
}

export interface BERTH_OFFER_WITHOUT_APPLICATION_PROFILE_profile {
  __typename: "ProfileNode";
  id: string;
  boats: BERTH_OFFER_WITHOUT_APPLICATION_PROFILE_profile_boats | null;
}

export interface BERTH_OFFER_WITHOUT_APPLICATION_PROFILE {
  profile: BERTH_OFFER_WITHOUT_APPLICATION_PROFILE_profile | null;
}

export interface BERTH_OFFER_WITHOUT_APPLICATION_PROFILEVariables {
  customerId: string;
}
