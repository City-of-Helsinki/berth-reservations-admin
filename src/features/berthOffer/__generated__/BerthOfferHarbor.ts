/* tslint:disable */
/* eslint-disable */
// @generated
// This file was automatically generated and should not be edited.

// ====================================================
// GraphQL fragment: BerthOfferHarbor
// ====================================================

export interface BerthOfferHarbor_properties_maps {
  __typename: "HarborMapType";
  id: any;
  url: string;
}

export interface BerthOfferHarbor_properties {
  __typename: "HarborProperties";
  name: string | null;
  servicemapId: string | null;
  imageFile: string | null;
  maps: (BerthOfferHarbor_properties_maps | null)[];
  streetAddress: string | null;
  municipality: string | null;
  zipCode: string;
  maxWidth: number | null;
  numberOfPlaces: number;
  numberOfFreePlaces: number;
}

export interface BerthOfferHarbor {
  __typename: "HarborNode";
  id: string;
  properties: BerthOfferHarbor_properties | null;
}
