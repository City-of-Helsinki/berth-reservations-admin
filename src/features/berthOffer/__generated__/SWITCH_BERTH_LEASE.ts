/* tslint:disable */
/* eslint-disable */
// @generated
// This file was automatically generated and should not be edited.

// ====================================================
// GraphQL query operation: SWITCH_BERTH_LEASE
// ====================================================

export interface SWITCH_BERTH_LEASE_berthLease_boat_boatType {
  __typename: "BoatTypeType";
  id: string;
  name: string | null;
}

export interface SWITCH_BERTH_LEASE_berthLease_boat {
  __typename: "BoatNode";
  boatType: SWITCH_BERTH_LEASE_berthLease_boat_boatType;
  draught: any | null;
  id: string;
  length: any;
  model: string;
  name: string;
  registrationNumber: string;
  weight: number | null;
  width: any;
}

export interface SWITCH_BERTH_LEASE_berthLease {
  __typename: "BerthLeaseNode";
  id: string;
  boat: SWITCH_BERTH_LEASE_berthLease_boat | null;
}

export interface SWITCH_BERTH_LEASE {
  berthLease: SWITCH_BERTH_LEASE_berthLease | null;
}

export interface SWITCH_BERTH_LEASEVariables {
  leaseId: string;
}
