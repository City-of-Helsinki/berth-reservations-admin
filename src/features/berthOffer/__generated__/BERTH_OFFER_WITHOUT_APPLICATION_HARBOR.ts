/* tslint:disable */
/* eslint-disable */
// @generated
// This file was automatically generated and should not be edited.

import { LeaseStatus, BerthMooringType } from "./../../../@types/__generated__/globalTypes";

// ====================================================
// GraphQL query operation: BERTH_OFFER_WITHOUT_APPLICATION_HARBOR
// ====================================================

export interface BERTH_OFFER_WITHOUT_APPLICATION_HARBOR_harbor_properties_piers_edges_node_properties_berths_edges_node_leases_edges_node_customer {
  __typename: "ProfileNode";
  id: string;
}

export interface BERTH_OFFER_WITHOUT_APPLICATION_HARBOR_harbor_properties_piers_edges_node_properties_berths_edges_node_leases_edges_node {
  __typename: "BerthLeaseNode";
  status: LeaseStatus;
  startDate: any;
  endDate: any;
  isActive: boolean;
  customer: BERTH_OFFER_WITHOUT_APPLICATION_HARBOR_harbor_properties_piers_edges_node_properties_berths_edges_node_leases_edges_node_customer;
}

export interface BERTH_OFFER_WITHOUT_APPLICATION_HARBOR_harbor_properties_piers_edges_node_properties_berths_edges_node_leases_edges {
  __typename: "BerthLeaseNodeEdge";
  node: BERTH_OFFER_WITHOUT_APPLICATION_HARBOR_harbor_properties_piers_edges_node_properties_berths_edges_node_leases_edges_node | null;
}

export interface BERTH_OFFER_WITHOUT_APPLICATION_HARBOR_harbor_properties_piers_edges_node_properties_berths_edges_node_leases {
  __typename: "BerthLeaseNodeConnection";
  edges: (BERTH_OFFER_WITHOUT_APPLICATION_HARBOR_harbor_properties_piers_edges_node_properties_berths_edges_node_leases_edges | null)[];
}

export interface BERTH_OFFER_WITHOUT_APPLICATION_HARBOR_harbor_properties_piers_edges_node_properties_berths_edges_node_pendingSwitchOffer_customer {
  __typename: "ProfileNode";
  id: string;
}

export interface BERTH_OFFER_WITHOUT_APPLICATION_HARBOR_harbor_properties_piers_edges_node_properties_berths_edges_node_pendingSwitchOffer {
  __typename: "BerthSwitchOfferNode";
  customer: BERTH_OFFER_WITHOUT_APPLICATION_HARBOR_harbor_properties_piers_edges_node_properties_berths_edges_node_pendingSwitchOffer_customer;
}

export interface BERTH_OFFER_WITHOUT_APPLICATION_HARBOR_harbor_properties_piers_edges_node_properties_berths_edges_node {
  __typename: "BerthNode";
  comment: string;
  depth: number | null;
  id: string;
  isAccessible: boolean | null;
  isActive: boolean;
  leases: BERTH_OFFER_WITHOUT_APPLICATION_HARBOR_harbor_properties_piers_edges_node_properties_berths_edges_node_leases | null;
  pendingSwitchOffer: BERTH_OFFER_WITHOUT_APPLICATION_HARBOR_harbor_properties_piers_edges_node_properties_berths_edges_node_pendingSwitchOffer | null;
  length: number;
  mooringType: BerthMooringType;
  number: string;
  width: number;
}

export interface BERTH_OFFER_WITHOUT_APPLICATION_HARBOR_harbor_properties_piers_edges_node_properties_berths_edges {
  __typename: "BerthNodeEdge";
  node: BERTH_OFFER_WITHOUT_APPLICATION_HARBOR_harbor_properties_piers_edges_node_properties_berths_edges_node | null;
}

export interface BERTH_OFFER_WITHOUT_APPLICATION_HARBOR_harbor_properties_piers_edges_node_properties_berths {
  __typename: "BerthNodeConnection";
  edges: (BERTH_OFFER_WITHOUT_APPLICATION_HARBOR_harbor_properties_piers_edges_node_properties_berths_edges | null)[];
}

export interface BERTH_OFFER_WITHOUT_APPLICATION_HARBOR_harbor_properties_piers_edges_node_properties {
  __typename: "PierProperties";
  identifier: string;
  electricity: boolean;
  gate: boolean;
  water: boolean;
  lighting: boolean;
  wasteCollection: boolean;
  berths: BERTH_OFFER_WITHOUT_APPLICATION_HARBOR_harbor_properties_piers_edges_node_properties_berths;
}

export interface BERTH_OFFER_WITHOUT_APPLICATION_HARBOR_harbor_properties_piers_edges_node {
  __typename: "PierNode";
  id: string;
  properties: BERTH_OFFER_WITHOUT_APPLICATION_HARBOR_harbor_properties_piers_edges_node_properties | null;
}

export interface BERTH_OFFER_WITHOUT_APPLICATION_HARBOR_harbor_properties_piers_edges {
  __typename: "PierNodeEdge";
  node: BERTH_OFFER_WITHOUT_APPLICATION_HARBOR_harbor_properties_piers_edges_node | null;
}

export interface BERTH_OFFER_WITHOUT_APPLICATION_HARBOR_harbor_properties_piers {
  __typename: "PierNodeConnection";
  edges: (BERTH_OFFER_WITHOUT_APPLICATION_HARBOR_harbor_properties_piers_edges | null)[];
}

export interface BERTH_OFFER_WITHOUT_APPLICATION_HARBOR_harbor_properties {
  __typename: "HarborProperties";
  name: string | null;
  servicemapId: string | null;
  imageFile: string | null;
  streetAddress: string | null;
  municipality: string | null;
  zipCode: string;
  maxWidth: number | null;
  numberOfPlaces: number;
  numberOfFreePlaces: number;
  piers: BERTH_OFFER_WITHOUT_APPLICATION_HARBOR_harbor_properties_piers | null;
}

export interface BERTH_OFFER_WITHOUT_APPLICATION_HARBOR_harbor {
  __typename: "HarborNode";
  id: string;
  properties: BERTH_OFFER_WITHOUT_APPLICATION_HARBOR_harbor_properties | null;
}

export interface BERTH_OFFER_WITHOUT_APPLICATION_HARBOR {
  harbor: BERTH_OFFER_WITHOUT_APPLICATION_HARBOR_harbor | null;
}

export interface BERTH_OFFER_WITHOUT_APPLICATION_HARBORVariables {
  harborId: string;
  boatWidth: number;
}
