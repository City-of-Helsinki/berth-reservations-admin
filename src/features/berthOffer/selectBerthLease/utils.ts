import { SELECT_BERTH_LEASE } from './__generated__/SELECT_BERTH_LEASE';
import { BerthLease } from './types';

export const getBerthLeases = (data: SELECT_BERTH_LEASE | undefined): BerthLease[] => {
  if (!data?.profile?.berthLeases) return [];
  return data.profile.berthLeases.edges.reduce<BerthLease[]>((acc, edge) => {
    if (!edge?.node?.isActive) return acc;
    const lease = edge.node;

    const harbor = lease.berth.pier.properties?.harbor.properties?.name ?? '-';
    const pier = lease.berth.pier.properties?.identifier ?? '-';
    const berth = lease.berth.number ?? '-';

    return [
      ...acc,
      {
        id: lease.id,
        name: [harbor, pier, berth].join(' '),
      },
    ];
  }, []);
};
