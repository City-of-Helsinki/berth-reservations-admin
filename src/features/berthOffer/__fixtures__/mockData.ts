import { BERTH_OFFER } from '../__generated__/BERTH_OFFER';
import { ApplicationStatus, BerthMooringType, LeaseStatus } from '../../../@types/__generated__/globalTypes';
import { SWITCH_BERTH_LEASE } from '../__generated__/SWITCH_BERTH_LEASE';

export const berthOfferQueryData: BERTH_OFFER = {
  berthApplication: {
    __typename: 'BerthApplicationNode',
    id: '1',
    createdAt: null,
    status: ApplicationStatus.PENDING,
    berthSwitch: null,
    customer: {
      __typename: 'ProfileNode',
      id: '1',
    },
    boatType: 'rowboat',
    boatRegistrationNumber: '1234',
    boatName: 'test',
    boatModel: 'test',
    boatWidth: 1,
    boatLength: 1,
    boatDraught: 1,
    boatWeight: 1,
  },
  boatTypes: [
    {
      __typename: 'BoatTypeType',
      id: 'rowboat',
      name: 'rowboat',
    },
  ],
  harbor: {
    __typename: 'HarborNode',
    id: 'SGFyYm9yTm9kZTpmN2M2YTQwZjAtOWViMi0zZjgyMTI0YjY0OGI=',
    properties: {
      __typename: 'HarborProperties',
      name: 'Pajalahden venesatama (Meripuistotie) / Venesatama',
      streetAddress: 'Meripuistotie 1a',
      zipCode: '00210',
      municipality: 'Helsinki',
      imageFile: 'https://venepaikat-api.test.hel.ninja/media/harbors/9a8d8313-eaa2-47d2-8f2d-2bb9893f9bc7/41359.jpg',
      maxWidth: 4,
      servicemapId: '41359',
      numberOfPlaces: 4,
      numberOfFreePlaces: 2,
      piers: {
        __typename: 'PierNodeConnection',
        edges: [
          {
            __typename: 'PierNodeEdge',
            node: {
              __typename: 'PierNode',
              id: '123456',
              properties: {
                __typename: 'PierProperties',
                identifier: 'pier',
                electricity: true,
                gate: true,
                lighting: true,
                wasteCollection: true,
                water: true,
                berths: {
                  __typename: 'BerthNodeConnection',
                  edges: [
                    {
                      __typename: 'BerthNodeEdge',
                      node: {
                        __typename: 'BerthNode',
                        comment: 'No comments',
                        depth: 5,
                        id: '1',
                        isAccessible: true,
                        isActive: true,
                        leases: {
                          __typename: 'BerthLeaseNodeConnection',
                          edges: [
                            {
                              __typename: 'BerthLeaseNodeEdge',
                              node: {
                                __typename: 'BerthLeaseNode',
                                status: LeaseStatus.PAID,
                                startDate: null,
                                endDate: null,
                                isActive: true,
                                customer: {
                                  __typename: 'ProfileNode',
                                  id: '1',
                                },
                              },
                            },
                          ],
                        },
                        length: 10,
                        mooringType: BerthMooringType.DINGHY_PLACE,
                        number: '1',
                        width: 5,
                      },
                    },
                  ],
                },
              },
            },
          },
        ],
      },
    },
  },
};

export const mockSwitchBerthLease: SWITCH_BERTH_LEASE = {
  berthLease: {
    __typename: 'BerthLeaseNode',
    boat: {
      __typename: 'BoatNode',
      boatType: { id: '2', name: 'Soutuvene', __typename: 'BoatTypeType' },
      draught: 0.5,
      id: 'MOCK-BOAT',
      length: 3,
      model: 'MODEL',
      name: 'BOAT',
      registrationNumber: '',
      weight: 100,
      width: 1,
    },
    id: 'MOCK-LEASE',
  },
};
