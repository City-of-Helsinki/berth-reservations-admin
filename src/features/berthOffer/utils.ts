import { BERTH_OFFER } from './__generated__/BERTH_OFFER';
import { LeaseStatus } from '../../@types/__generated__/globalTypes';
import { HarborCardProps } from '../../common/harborCard/HarborCard';
import { Boat } from '../../common/boatCard/types';
import { HarborData, BerthData, PiersData, BoatData, Lease, PierTab } from './types';
import { SWITCH_BERTH_LEASE } from './__generated__/SWITCH_BERTH_LEASE';

export const getBerthData = (data: HarborData | null | undefined): BerthData[] => {
  if (!data?.properties?.piers) return [];

  const harborId = data.id;
  const harbor = data.properties.name || '';
  return data.properties.piers.edges.reduce<BerthData[]>((acc, pier) => {
    if (!pier?.node?.properties) return acc;

    const { properties } = pier.node;
    const berths = pier.node.properties.berths.edges.reduce<BerthData[]>((acc, berth) => {
      if (!berth?.node) return acc;

      const leases =
        berth.node.leases?.edges.reduce<Lease[]>((acc, edge) => {
          if (!edge?.node || edge?.node?.status !== LeaseStatus.PAID) return acc;

          return [
            ...acc,
            {
              startDate: edge.node.startDate,
              endDate: edge.node.endDate,
              status: edge.node.status,
              isActive: edge.node.isActive,
              customer: {
                id: edge.node.customer.id,
              },
            },
          ];
        }, []) ?? [];

      return [
        ...acc,
        {
          berth: berth.node.number,
          berthId: berth.node.id,
          comment: berth.node.comment,
          draught: berth.node.depth,
          harbor,
          harborId,
          id: berth.node.id,
          isActive: berth.node.isActive,
          leases,
          length: berth.node.length,
          mooringType: berth.node.mooringType,
          pier: properties.identifier,
          properties: {
            lighting: properties.lighting,
            water: properties.water,
            gate: properties.gate,
            electricity: properties.electricity,
            wasteCollection: properties.wasteCollection,
            isAccessible: berth.node.isAccessible,
          },
          width: berth.node.width,
          pendingSwitchOffer: berth.node.pendingSwitchOffer,
        },
      ];
    }, []);

    return [...acc, ...berths];
  }, []);
};

export const getAllPiersIdentifiers = (data: PiersData | null | undefined): PierTab[] => {
  const piers = data?.edges ?? [];

  return piers.reduce<PierTab[]>((acc, pier) => {
    if (!pier?.node?.properties) return acc;

    const pierTab = {
      label: pier.node.properties.identifier,
      value: pier.node.properties.identifier,
      disabled: !pier.node.properties?.berths.edges.length,
    };

    return [...acc, pierTab];
  }, []);
};

export const getHarbor = (data: BERTH_OFFER['harbor'] | undefined): HarborCardProps['harbor'] | null => {
  if (!data?.properties?.piers) {
    return null;
  }

  const pierProps = data.properties.piers.edges.reduce(
    (prev, pier) => {
      if (pier?.node?.properties) {
        return {
          electricity: prev.electricity || pier.node.properties.electricity,
          gate: prev.gate || pier.node.properties.gate,
          lighting: prev.lighting || pier.node.properties.lighting,
          wasteCollection: prev.wasteCollection || pier.node.properties.wasteCollection,
          water: prev.water || pier.node.properties.water,
        };
      }
      return prev;
    },
    {
      electricity: false,
      gate: false,
      lighting: false,
      wasteCollection: false,
      water: false,
    }
  );

  const { properties } = data;
  return {
    imageUrl: properties.imageFile,
    name: properties.name || '',
    streetAddress: properties.streetAddress,
    municipality: properties.municipality,
    zipCode: properties.zipCode,
    servicemapId: properties.servicemapId || '',
    properties: {
      ...pierProps,
      queue: 0, // TODO
      numberOfPlaces: properties.numberOfPlaces,
      numberOfFreePlaces: properties.numberOfFreePlaces,
      maxWidth: properties.maxWidth || 0,
    },
  };
};

export const getApplicationBoat = (
  application: BERTH_OFFER['berthApplication'] | undefined,
  boatTypes: BERTH_OFFER['boatTypes'] | undefined
): Boat | null => {
  if (!application) return null;

  return {
    boatType: boatTypes?.find(({ id }) => id === application.boatType)?.name ?? null,
    boatRegistrationNumber: application.boatRegistrationNumber,
    boatName: application.boatName,
    boatModel: application.boatModel,
    boatWidth: application.boatWidth,
    boatLength: application.boatLength,
    boatDraught: application.boatDraught,
    boatWeight: application.boatWeight,
  };
};

export const getLeaseBoat = (data: SWITCH_BERTH_LEASE | undefined): Boat | undefined => {
  const boat = data?.berthLease?.boat;
  if (!boat) return undefined;

  return {
    boatRegistrationNumber: boat.registrationNumber,
    boatType: boat.boatType.name,
    boatName: boat.name,
    boatWidth: boat.width ? Number(boat.width) : boat.width,
    boatDraught: boat.draught ? Number(boat.draught) : boat.draught,
    boatLength: boat.length ? Number(boat.length) : boat.length,
    boatWeight: boat.weight ? Number(boat.weight) : boat.weight,
    boatModel: boat.model,
  };
};

export const getCustomerBoat = (data: BoatData | null | undefined): Boat | null => {
  if (!data) return null;

  return {
    boatRegistrationNumber: data.registrationNumber,
    boatType: data.boatType.name,
    boatName: data.name,
    boatWidth: data.width ? Number(data.width) : data.width,
    boatDraught: data.draught ? Number(data.draught) : data.draught,
    boatLength: data.length ? Number(data.length) : data.length,
    boatWeight: data.weight ? Number(data.weight) : data.weight,
    boatModel: data.model,
  };
};

export const isSuitableBerthLength = (berthLength: number, boatLength: number) => berthLength >= boatLength;

export const getApplicationTypeTKey = (isSwitch: boolean): string =>
  isSwitch ? 'applicationList.applicationType.switchApplication' : 'applicationList.applicationType.newApplication';
