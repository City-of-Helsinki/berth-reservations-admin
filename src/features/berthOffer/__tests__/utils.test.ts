import { getAllPiersIdentifiers, getHarbor, getBerthData, getLeaseBoat } from '../utils';
import { berthOfferQueryData, mockSwitchBerthLease } from '../__fixtures__/mockData';
import { SWITCH_BERTH_LEASE } from '../__generated__/SWITCH_BERTH_LEASE';

describe('utils', () => {
  describe('getOfferData', () => {
    it('should return offer data', () => {
      const berthData = getBerthData(berthOfferQueryData['harbor']);

      expect(berthData).toMatchSnapshot();
    });
  });

  describe('getAllPiersIdentifiers', () => {
    it('should return pier tabs', () => {
      const harbor = getAllPiersIdentifiers(berthOfferQueryData['harbor']?.properties?.piers);

      expect(harbor).toMatchSnapshot();
    });
  });

  describe('getHarbor', () => {
    it('should return null when the provided data is undefined', () => {
      expect(getHarbor(undefined)).toStrictEqual(null);
    });

    it('should return harbor', () => {
      const harbor = getHarbor(berthOfferQueryData.harbor);

      expect(harbor).toMatchSnapshot();
    });
  });

  describe('getLeaseBoat', () => {
    it('should map the lease boat correctly', () => {
      expect(getLeaseBoat(mockSwitchBerthLease)).toEqual({
        boatDraught: 0.5,
        boatLength: 3,
        boatModel: 'MODEL',
        boatName: 'BOAT',
        boatRegistrationNumber: '',
        boatType: 'Soutuvene',
        boatWeight: 100,
        boatWidth: 1,
      });
    });

    it('should return undefined is no boat exists', () => {
      const noBoat: SWITCH_BERTH_LEASE = {
        berthLease: {
          __typename: 'BerthLeaseNode',
          boat: null,
          id: 'MOCK-LEASE',
        },
      };

      expect(getLeaseBoat(noBoat)).toEqual(undefined);
    });
  });
});
