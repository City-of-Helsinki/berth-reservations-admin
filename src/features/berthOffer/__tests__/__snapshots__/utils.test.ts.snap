// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`utils getAllPiersIdentifiers should return pier tabs 1`] = `
Array [
  Object {
    "disabled": false,
    "label": "pier",
    "value": "pier",
  },
]
`;

exports[`utils getHarbor should return harbor 1`] = `
Object {
  "imageUrl": "https://venepaikat-api.test.hel.ninja/media/harbors/9a8d8313-eaa2-47d2-8f2d-2bb9893f9bc7/41359.jpg",
  "municipality": "Helsinki",
  "name": "Pajalahden venesatama (Meripuistotie) / Venesatama",
  "properties": Object {
    "electricity": true,
    "gate": true,
    "lighting": true,
    "maxWidth": 4,
    "numberOfFreePlaces": 2,
    "numberOfPlaces": 4,
    "queue": 0,
    "wasteCollection": true,
    "water": true,
  },
  "servicemapId": "41359",
  "streetAddress": "Meripuistotie 1a",
  "zipCode": "00210",
}
`;

exports[`utils getOfferData should return offer data 1`] = `
Array [
  Object {
    "berth": "1",
    "berthId": "1",
    "comment": "No comments",
    "draught": 5,
    "harbor": "Pajalahden venesatama (Meripuistotie) / Venesatama",
    "harborId": "SGFyYm9yTm9kZTpmN2M2YTQwZjAtOWViMi0zZjgyMTI0YjY0OGI=",
    "id": "1",
    "isActive": true,
    "leases": Array [
      Object {
        "customer": Object {
          "id": "1",
        },
        "endDate": null,
        "isActive": true,
        "startDate": null,
        "status": "PAID",
      },
    ],
    "length": 10,
    "mooringType": "DINGHY_PLACE",
    "pier": "pier",
    "properties": Object {
      "electricity": true,
      "gate": true,
      "isAccessible": true,
      "lighting": true,
      "wasteCollection": true,
      "water": true,
    },
    "width": 5,
  },
]
`;
